### PR TITLE
Pal 525 service actuator endpoints

### PR DESCRIPTION
**Pull request by developer01189998819991197253**

### CODE REVIEW

File: data-service/src/main/resources/application-local.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+web:
+  client:
+    palisade-service: "localhost:8084"
```
Comment (dev930018) [inline data-service/src/main/resources/application-local.yaml]: There should now be suitable eureka integration that the web client fields are unrequired.
By default in other services, the field has been left as `palisade-service: palisade-service`

File: data-service/src/main/java/uk/gov/gchq/palisade/service/data/web/PalisadeClient.java [hunk 2]
```diff
@@ -30,4 +32,7 @@
     @PostMapping(path = "/getDataRequestConfig", consumes = "application/json", produces = "application/json")
     DataRequestConfig getDataRequestConfig(final URI url, @RequestBody final GetDataRequestConfig request);
 
+    @GetMapping(path = "/actuator/health", produces = "application/*+json")
+    Response getHealth();
```
Comment (dev930018) [inline data-service/src/main/java/uk/gov/gchq/palisade/service/data/web/PalisadeClient.java]: Could this be changed to use eureka (using the URI first argument)?

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/web/PolicyClient.java [hunk 3]
```diff
@@ -32,4 +34,7 @@
     @PostMapping(path = "/getPolicySync", consumes = "application/json", produces = "application/json")
     Map<LeafResource, Rules> getPolicySync(final URI url, @RequestBody final GetPolicyRequest request);
 
+    @GetMapping(path = "/actuator/health", produces = "application/*+json")
+    Response getHealth();
```
Comment (dev930018) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/web/PolicyClient.java]: Ditto, use URI for eureka

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/web/ResourceClient.java [hunk 4]
```diff
@@ -48,4 +50,7 @@
     @PostMapping(path = "/addResource", consumes = "application/json", produces = "application/json")
     Boolean addResource(final URI url, @RequestBody final AddResourceRequest request);
 
+    @GetMapping(path = "/actuator/health", produces = "application/json")
+    Response getHealth();
```
Comment (dev930018) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/web/ResourceClient.java]: URI for eureka

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/web/UserClient.java [hunk 5]
```diff
@@ -30,4 +32,7 @@
     @PostMapping(path = "/getUser", consumes = "application/json", produces = "application/json")
     User getUser(final URI url, @RequestBody final GetUserRequest request);
 
+    @GetMapping(path = "/actuator/health", produces = "application/json")
+    Response getHealth();
```
Comment (dev930018) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/web/UserClient.java]: URI for eureka

File: palisade-service/src/main/resources/application-local.yaml [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+web:
+  client:
+    audit-service: "localhost:8081"
```
Comment (dev930018) [inline palisade-service/src/main/resources/application-local.yaml]: Ditto, these fields should no longer be required using eureka

File: services-manager/src/main/resources/application-services.yaml [hunk 7]
```diff
@@ -46,6 +46,7 @@ manager:
       paths: *libraries
       profiles:
         - eureka
+        - local
```
Comment (dev930018) [inline services-manager/src/main/resources/application-services.yaml]: Thsese profiles shouldn't be needed with eureka

File: services-manager/src/main/resources/application-services.yaml [hunk 8]
```diff
@@ -55,6 +56,7 @@ manager:
       profiles:
         - eureka
         - jvm
+        - local
```
Comment (dev930018) [inline services-manager/src/main/resources/application-services.yaml]: ditto

**Review by dev930018 — CHANGES_REQUESTED** [review]
Could you add eureka integration?
For an example, see (all in the *data-service*):
* config/ClientConfiguration.java
* service/PalisadeService.java
* web/PalisadeClient.java
* resources/application.yaml (and all others, there should no longer be a need for *localhost:PORT*)

**Comment by dev930018** [comment]
The above will all also need to be done (to a lesser extent) for the data service. Also, this misses out the audit service!
[x] palisade-service connects to user,resource,policy so is up when they are
[ ] data-service connects to palisade so is up when it is
[ ] palisade,data both connect to audit and fail if auditing is unreachable, can't be done etc. so both are up when audit is

tl;dr - we need the following dependencies in the health actuators:
* palisade -> user, resource, policy, *audit*
* data -> *palisade*, *audit*

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/web/PalisadeHealthIndicator.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        Response userHealth = userService.getHealth();
+        if (policyHealth.status() != 200) {
+            throw new Exception("Policy service down");
+        }
+        if (resourceHealth.status() != 200) {
+            throw new Exception("Resource service down");
+        }
+        if (userHealth.status() != 200) {
+            throw new Exception("User service down");
+        }
```
Comment (pd104923) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/web/PalisadeHealthIndicator.java:68]: Is there a reason we're not checking the status of auditHealth here?

File: data-service/src/main/resources/application-local.yaml (hunk 1 above)
Comment (developer01189998819991197253) [reply data-service/src/main/resources/application-local.yaml]: removed application-local to use eureka instead

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/web/ResourceClient.java [hunk 10]
```diff
@@ -48,4 +50,7 @@
     @PostMapping(path = "/addResource", consumes = "application/json", produces = "application/json")
     Boolean addResource(final URI url, @RequestBody final AddResourceRequest request);
 
+    @GetMapping(path = "/actuator/health", produces = "application/*+json")
```
Comment (dev930018) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/web/ResourceClient.java]: Why is this one json-like but the rest were json?

Comment (developer01189998819991197253) [reply palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/web/ResourceClient.java]: reverted

**Review by pd104923 — DISMISSED** [review]

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/config/ApplicationConfiguration.java [hunk 11]
```diff
@@ -141,6 +142,11 @@ public PolicyService policyService(final PolicyClient policyClient, final Client
         return new PolicyService(policyClient, policyUriSupplier, getAsyncExecutor());
     }
 
+    @Bean
+    public PalisadeHealthIndicator dataHealthIndicator(final AuditService auditService, final PolicyService policyService, final ResourceService resourceService, final UserService userService) {
```
Comment (dev930018) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/config/ApplicationConfiguration.java]: dataHealthIndicator? Meant to be palisadeHealthIndicator?

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/web/PalisadeHealthIndicator.java (hunk 9 above)
Comment (dev930018) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/web/PalisadeHealthIndicator.java:68]: This doesn't look to be checking the audit-service health still

**Review by dev930018 — CHANGES_REQUESTED** [review]

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/config/ApplicationConfiguration.java (hunk 11 above)
Comment (developer01189998819991197253) [reply palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/config/ApplicationConfiguration.java]: renamed

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/web/PalisadeHealthIndicator.java (hunk 9 above)
Comment (developer01189998819991197253) [reply palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/web/PalisadeHealthIndicator.java:68]: added

**Review by dev930018 — APPROVED** [review]

**Review by ac74475 — APPROVED** [review]
